renderer_vulkan: Handle more miscellaneous GPU settings

raphaelthegreat opened the pull request:
While wondering radv_cmd_buffer.c noticed there are a few things that were not handled at all, or handled incorrectly. This PR is a collection of some easier ones to implement.

The most complicated one is the depth clamp/clip commit. The existing code would clamp the viewport minDepth/maxDepth but that is actually incorrect as those affect the transformation of the Z values. The new depth clamp control extension gives us the ability to specify a custom depth clamp range, aside from the default behavior or using viewport setting.

raphaelthegreat committed:
renderer_vulkan: Respect provoking vertex setting

raphaelthegreat committed:
renderer_vulkan: Handle rasterization discard

raphaelthegreat committed:
renderer_vulkan: Implement logic ops

raphaelthegreat committed:
renderer_vulkan: Properly implement depth clamp and clip

raphaelthegreat committed:
renderer_vulkan: Handle line width

raphaelthegreat committed:
Fix build

raphaelthegreat committed:
vk_pipeline_cache: Don't check depth clamp without a depth buffer

squidbus commented on src/video_core/renderer_vulkan/vk_pipeline_cache.cpp:
```diff
@@ -491,7 +486,69 @@ bool PipelineCache::RefreshGraphicsKey() {
+            Render_Vulkan,
+            "Viewport uses custom clamp range zmin={}, zmax={} which cannot be accurately emulated",
+            zmin, zmax);
+    }
```
Do we need these warnings when depth clamp is disabled?

squidbus replied:
Also want to add that when the extension here is not supported multiple games will just completely fill the log with this and make it impossible to read.

raphaelthegreat replied:
Has this been observed with new commit that skips this when no depth buffer? So far I've only seen this warning in depth decompression draws with no framebuffers attached

squidbus replied:
Yes, I get a lot of `Viewport uses custom clamp range zmin=0.0, zmax=1.0 which cannot be accurately emulated` in certain games.

One of those games for example uses a viewport min-max depth range of 0.5-1.0 frequently (clip space = 0-W, offset = 0.5, scale = 0.5).

And Bloodborne spams this because it keeps its `zoffset = 0.0, zscale = 1.0` when switching to `MinusWToW` clip space, resulting in `min_depth = -1.0, max_depth = 1.0` when clamp is `zmin = 0.0, zmax = 1.0`.

raphaelthegreat replied:
I'm curious, can you check if this is an actual draw with depth testing or an empty one? I will disable the log though

raphaelthegreat replied:
Tweaked the viewport check to also allow cases where viewport depth range is more restrictive then depth clamp range, viewport range can be used in that case too

squidbus replied:
Seems to be gone in most games, just have it hit in Bloodborne which I suppose is a valid case (edited details into above comment).

georgemoralis replied:
having a pr that doesn't affect bb seems rare these days

squidbus replied:
Well to be clear it will still render the same, it just will output this log a lot. Which I guess isn’t a big issue considering that game already spams HTTP stub errors.

raphaelthegreat replied:
I can make it log on the graphics pipeline constructor to avoid spam

raphaelthegreat committed:
liverpool: Fix line control offset

raphaelthegreat committed:
vk_pipeline_cache: Don't run search if depth clamp is disabled

raphaelthegreat committed:
vk_pipeline_cache: Allow using viewport range when it's more restrictive then depth clamp

raphaelthegreat committed:
liverpool: Disable depth clip when near and far planes have different setting

raphaelthegreat committed:
vk_graphics_pipeline: Move warning to pipeline

raphaelthegreat committed:
vk_pipeline_cache: Revert viewport check and remove log

raphaelthegreat committed:
vk_graphics_pipeline: Enable depth clamp when depth clip is disabled and extension is not supported

Without the depth clip extension depth clipping is controlled by depth clamping
